### ISSUE TITLE
API: changing qos returns 403 but succeeds anyway

### ISSUE THREAD

**Issue by onnozweers** [dCache/dcache#5301]
Dear dCache devs,

I have a case in which a change in qos returns a 403, but the qos is changed anyway.

Here's a listing:

```
$ curl -H 'Authorization: Bearer <token>' -H 'accept: application/json' --fail --silent --show-error -X GET 'https://dolphin12.grid.surfsara.nl:20443/api/v1/namespace/%2Ftape%2FdCache.jpg?children=false&locality=true&locations=false&qos=true' | jq .
{
  "fileMimeType": "image/jpeg",
  "fileLocality": "ONLINE_AND_NEARLINE",
  "currentQos": "tape",
  "fileType": "REGULAR",
  "pnfsId": "0000F20EC86EA0484875848ED18B9D910888",
  "nlink": 1,
  "mtime": 1580898648206,
  "creationTime": 1580898647574,
  "size": 11099
}
```
Then I change the qos to `disk+tape`:
```
$ curl -H 'Authorization: Bearer <token>' -H 'accept: application/json' --fail --silent --show-error -H 'content-type: application/json' -X POST https://dolphin12.grid.surfsara.nl:20443/api/v1/namespace/%2Ftape%2FdCache.jpg -d '{"action":"qos","target":"disk+tape"}'
curl: (22) The requested URL returned error: 403 Forbidden
```
I do the listing again:
```
$ curl -H 'Authorization: Bearer <token>' -H 'accept: application/json' --fail --silent --show-error -X GET 'https://dolphin12.grid.surfsara.nl:20443/api/v1/namespace/%2Ftape%2FdCache.jpg?children=false&locality=true&locations=false&qos=true' | jq .
{
  "fileMimeType": "image/jpeg",
  "fileLocality": "ONLINE_AND_NEARLINE",
  "currentQos": "disk+tape",
  "fileType": "REGULAR",
  "pnfsId": "0000F20EC86EA0484875848ED18B9D910888",
  "nlink": 1,
  "mtime": 1580898648206,
  "creationTime": 1580898647574,
  "size": 11099
}
```
So, the transition has succeeded.

Interestingly, when I change the qos back to `tape`, it returns a `success`.
```
$ curl -H 'Authorization: Bearer <token>' -H 'accept: application/json' --fail --silent --show-error -H 'content-type: application/json' -X POST https://dolphin12.grid.surfsara.nl:20443/api/v1/namespace/%2Ftape%2FdCache.jpg -d '{"action":"qos","target":"tape"}'
{"status":"success"}
```

Tested in 6.1.0.9138f84.

Cheers,
Onno

**Comment by alrossi** [comment]
Onno, I just started investigating this, but it looks to me like this is a bug not in QoS but in the authorization mechanism (bearer token).   I tried to reproduce this with credential authentication and could not get the 403 error code.

**Comment by onnozweers** [comment]
Hi Al, thanks for looking into this.

It happens when I use a macaroon with a `cid root:/users/onno/` caveat. When I have a macaroon with a `cid path:/users/onno/` caveat, it works as expected.

Cheers,
Onno

**Comment by alrossi** [comment]
Ah, that's useful information, thanks.

**Comment by femiadeyemi** [comment]
You are correct @alrossi, this is macaroons related issue. As far as I see it, I don't think this is a _bug_ per se but I'm not 100% sure. Anyway, I think @paulmillar will be able to provide more insight.

**Comment by femiadeyemi** [comment]
Here is another related issue: https://github.com/dCache/dcache/issues/5318

**Comment by paulmillar** [comment]
Indeed.  Given removing the `root` caveat allows the command to succeed, it strongly looks like like the problem is frontend is not calculating the correct path.

Additionally, it is worrying that an operation that reports to the client that it failed should end up succeeding.  Depending on the actual cause, this could be an additional bug.

The underlying cause is (very likely) the same as in #5318: the dCache-internal path is calculated directly, without taking the `root` caveat into account.  Given we have had two reports of this issue, there are likely more places where this problem is present.

**Comment by alrossi** [comment]
Closing because it was solved (apparently) by #5318
